tutorials/multicore/mt301_fillHistAutoBin.C

```diff
@@ -0,0 +1,95 @@
+/// \file
+/// \ingroup tutorial_multicore
+/// Fill histograms w/o in parallel with automatic binning.
+/// Illustrates range synchronization
+///
+/// \macro_code
+///
+/// \author Gerardo Ganis
+/// \date August 2017
+#include "TH1D.h"
+#include "TH2D.h"
+#include "TH3D.h"
+#include "TList.h"
+#include "TRandom3.h"
+#include "TDirectory.h"
+#include "TROOT.h"
+#include "TCanvas.h"
+#include <thread>
+#include <iostream>
+#include "ROOT/TSeq.hxx"
+#include "ROOT/TThreadedObject.hxx"
+
+// The number of workers
+const UInt_t nWorkers = 8U;
+
+Int_t mt301_fillHistos(UInt_t nNumbers = 1001, Bool_t ranges = kFALSE)
+{
+
+   // The first, fundamental operation to be performed in order to make ROOT
+   // thread-aware.
+   ROOT::EnableThreadSafety();
+
+   Double_t mi = 0., ma = -1.;
+   if (ranges) {
+      mi = -4.;
+      ma = 4.;
+   }
+
+   // Histograms to be filled in parallel
+   ROOT::TThreadedObject<TH1D> h1d("h1d", "1D test histogram", 64, mi, ma);
+   ROOT::TThreadedObject<TH2D> h2d("h2d", "2D test histogram", 64, mi, ma, 64, mi, ma);
+   ROOT::TThreadedObject<TH3D> h3d("h3d", "3D test histogram", 64, mi, ma, 64, mi, ma, 64, mi, ma);
+
+   // We define our work item
+   auto workItem = [&](UInt_t workerID) {
+      // One generator, file and ntuple per worker
+      TRandom3 workerRndm(workerID); // Change the seed
+
+      auto wh1d = h1d.Get();
+      auto wh2d = h2d.Get();
+      auto wh3d = h3d.Get();
+
+      Double_t x, y;
+
+      for (UInt_t i = 0; i < nNumbers; ++i) {
+         wh1d->Fill(workerRndm.Gaus());
+         workerRndm.Rannor(x, y);
+         wh2d->Fill(x, y);
+         workerRndm.Rannor(x, y);
+         wh3d->Fill(x, y, workerRndm.Gaus());
+      }
+   };
+
+   // Create the collection which will hold the threads, our "pool"
+   std::vector<std::thread> workers;
+
+   // Fill the "pool" with workers
+   for (auto workerID : ROOT::TSeqI(nWorkers)) {
+      workers.emplace_back(workItem, workerID);
+   }
+
+   // Now join them
+   for (auto &&worker : workers)
+      worker.join();
+
+   // Merge
+   auto fh1d = h1d.Merge();
+   auto fh2d = h2d.Merge();
+   auto fh3d = h3d.Merge();
+
+   // Make the canvas
+   TCanvas *c = new TCanvas("c", "c", 800, 800);
+   c->Divide(2, 2);
+
+   c->cd(1);
+   fh1d->DrawClone();
+   c->cd(2);
+   fh2d->DrawClone();
+   c->cd(3);
+   fh3d->DrawClone();
+
+   gROOTMutex = 0;
```

Comment:
why this?

Reply:
This is a remnant to be able to run more times in the same ROOT shell.
I will remove it, also because all our mt... tutorials suffer from the same problem, i.e. at the second run all threads get locked at TRoot::Append:

 #0  __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
#1  0x00007fd50c469e42 in __GI___pthread_mutex_lock (mutex=0x3cd6060) at ../nptl/pthread_mutex_lock.c:115
#2  0x00007fd50b096f58 in TPosixMutex::Lock (this=0x3cd6050) at /home/ganis/local/root/GIT/root/core/thread/src/TPosixMutex.cxx:75
#3  0x00007fd50b0881cb in TMutex::Lock (this=0x3d117e0) at /home/ganis/local/root/GIT/root/core/thread/src/TMutex.cxx:48
#4  0x00007fd50e0412a2 in TLockGuard::TLockGuard (this=0x7fd4fcd73100, mutex=0x3d117e0) at include/TVirtualMutex.h:77
#5  0x00007fd50dac67fe in TROOT::Append (this=0x7fd50dff3240 <ROOT::Internal::GetROOT1()::alloc>, obj=0x7fd4e00008c0, replace=false)
    at /home/ganis/local/root/GIT/root/core/base/src/TROOT.cxx:1003
#6  0x00007fd501e1fca3 in TH1::Copy (this=0x48499e0, obj=...) at /home/ganis/local/root/GIT/root/hist/hist/src/TH1.cxx:2729
#7  0x00007fd501e3b82f in TH1F::Copy (this=0x48499e0, newth1=...) at /home/ganis/local/root/GIT/root/hist/hist/src/TH1.cxx:9426
#8  0x00007fd501e3b6ba in TH1F::TH1F (this=0x7fd4e00008c0, h=...) at /home/ganis/local/root/GIT/root/hist/hist/src/TH1.cxx:9411